### PR TITLE
Resolve #2415: Lucene: cross transactional merges - extra cleanup

### LINKED ISSUE #2415: Lucene: cross transactional merges - extra cleanup
* Move FIleLock code to FDBDirectoryLock
* FDBDirectoryLock: add test
* AgilityContext: add test

### PR DESCRIPTION
**Pull request by jjezra**
* Move FIleLock code to FDBDirectoryLock
  * FDBDirectoryLock: add test
  * AgilityContext: add test

### CODE REVIEW

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryTest.java
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryTest.java:265]: It would be good to have a test of the atomicity of `apply`/`accept`. You may have better ideas, but the first one that jumps into my mind is to have a bunch of threads that do something like:
```
context.apply(aContext -> {
    value = randomBytes()
    aContext.set(key1,  value);
    aContext.set(key2, value);
});
```
and a bunch of other threads doing:
```
context.accept(aContext -> {
    assertEquals(aContext.get(key2), aContext.get(key1)); // intentionally in the opposite order
});
```

Comment (ScottDugas) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryTest.java:265]: Probably also worth testing that the non-agile one's atomicity, namely that a series of `get`/`set` interspersed should not be seeing other contexts work.

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryTest.java
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryTest.java:290]: `get`/`set` are both very fast, would it improve the coverage of concurrency by using `apply`/`accept` that do a bit more? Or perhaps having it fast is better? Maybe both would be worthwhile.

**Review by ScottDugas — CHANGES_REQUESTED** [review]
I listed a bunch of additional tests that could be added, that could probably be deferred, as getting more coverage in now is better.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryLockFactory.java
Comment (jjezra) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryLockFactory.java:79]: IIUIC, Lucene is expecting a simple file as a lock. `SimpleFSLockFactory`, at least, does not use an FS lock on the file, hence `ensureValid()` is not expected to be atomic or to be performed under a lock.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryLockFactory.java
Comment (jjezra) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryLockFactory.java:127]: I believe so. Given the existing code, this is what Lucene is expecting and assuming.

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryTest.java
Comment (jjezra) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryTest.java:297]: Yes, but Intelij insisted.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryLockFactory.java
Comment (ScottDugas) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryLockFactory.java:127]: Could you please elaborate?
Creating and committing is not super cheap, so if we can avoid this commit that would be beneficial.

I don't think this should block this PR, but something worth investigating.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryLockFactory.java
Comment (ScottDugas) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryLockFactory.java:79]: AFAICT `SimpleFSLockFactory` will never steal the lock, so, I think, we have an additional risk here.
I think making it transactional would be ideal, if we can (after a brief scan of the code, that looks tricky at best), so perhaps having `ensureValid()` check that the timestamp is no more than 30 minutes ago? Which should help ensure that another instance won't steal before the next time it does an operation (assuming clock skew is under 30 minutes).

Comment (jjezra) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryLockFactory.java:79]: Do you mean the local timestamp? 
(I.e. `timeStampMillis`)
If so, why 30 minutes and not an hour (at least until we'll implement an heart bit)?

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryLockFactory.java
Comment (jjezra) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryLockFactory.java:127]: It is my understanding that if Lucene sets the file lock, it is expecting other processes to see it. Maybe preventing a double merge, maybe something else. 
Coming to think about it, in a parallel merge (or so), only one process will successfully commit the lock (eventually). So it's probably harmless to skip the flush. I'll do that.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryLockFactory.java
Comment (ScottDugas) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryLockFactory.java:79]: Yes, I mean `timeStampMillis`.
Using 30 minutes provides some buffer for clock skew, and GC pauses and the operations between calling validateLock.
So if you have two instances: a and b, and b has a clock skew putting it 1 minute ahead you'll have:
t0: a creates a lock at t0
t53: a validates the lock, and proceeds to do something that takes 8 minutes before validating again
t59: b (which believes it is t60) tries to create a lock, and concludes that the current lock at t0 is one hour in the past, and should be stolen.
t61: a validates and sees that it no-longer has the lock, but at this point it wrote some data after b stole the lock


I'm not sure exactly what these two numbers should be, but 30 minutes seems like more than enough time to accomplish a merge in the short term, and gives plenty of buffer to avoid continuing to do work after another process has stolen the lock.

Looking at the usages, it looks like it validates the lock whenever it creates files or modifies the file system, and each of those operations on a proper filesystem would get flushed to the filesystem (although not to disk) when the method is called, whereas we might create a "file" and then continue to write to the file and do other things before it is actually committed.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryLockFactory.java
Comment (ScottDugas) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryLockFactory.java:127]: Lucene is expecting other processes to see the lock, but the transactions should protect us here.
Namely, if you have two processes a and b, and
1. a gets a read version 10, creates the lock, and does some writes
2. b gets a read version 11, creates a lock, does some operations, then clears the lock and commits
3. a commits

then a will fail with a conflict, and nothing will be written.

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryTest.java
Comment (jjezra) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryTest.java:265]: ✅

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryLockFactory.java
Comment (jjezra) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryLockFactory.java:69]: ✅

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryLockFactory.java
Comment (jjezra) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryLockFactory.java:98]: Just additional sanity. I'll remove it.

Comment (jjezra) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryLockFactory.java:98]: ✅

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryLockFactory.java
Comment (jjezra) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryLockFactory.java:79]: ✅

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryLockFactory.java
Comment (jjezra) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryLockFactory.java:127]: Removed both flushes (at set and at clear). 
✅

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryTest.java
Comment (jjezra) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryTest.java:247]: ✅

Comment (jjezra) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryTest.java:247]: ✅

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryTest.java
Comment (jjezra) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryTest.java:272]: Created additional test for that. Also use multiple quota sizes. 
✅

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryTest.java
Comment (jjezra) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryTest.java:279]: ✅

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryTest.java
Comment (jjezra) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryTest.java:265]: ✅

Comment (jjezra) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryTest.java:265]: ✅

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryTest.java
Comment (jjezra) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryTest.java:290]: Implemented in the last test.

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java:136]: This quote is from a work still under copyright, it might fall under fair-use but it would be better to pick text that's definitely in the public domain (before 1900 should be ok)

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java:154]: Here also, you don't join the future and won't see failures.

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java:90]: These overwrite each other (e.g. loop=2 and thread=2 write the same value to the same key).
Is that intentional, or would it be better to have `Tuple.from(500, loop, thread)` and `Tuple.from(loop, thread)` ?

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java:183]: Perhaps would make more sense to set a field to a random bytestring in the setup, and use that for all tests, otherwise they will overlap, or scope all data to `FDBRecordStoreTestBase.path`, because `FDBRecordStoreTestBase` clears the range at the start of the test.

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java:182]:
```suggestion
    void testAgilityContextAtomicAttribute(boolean useAgile) {
        // test that commits won't happen in he middle of an accept or apply call
```

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java:194]: Would it make sense to put a 1ms sleep inside this thread?

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java:200]: We should also have a test where these are all done on different contexts, i.e. where each of these threads calls
```
try (FDBRecordContext context = openContext()) {
```
and uses that context.
It's important that a single agility context doesn't let reads and writes overlap, but I think even more important that it never commits in the middle of a `accept` or `apply`

In that version of the test, we would want to make sure that the size of the writes (or the limit) are such that the AgilityContext would choose to commit in the middle of the loop, if `commitIfNeeded` were called at that time.

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java:204]: I don't think it's particularly relevant for lucene, but we may, at some point, want to change `accept` to return a `CompletableFuture<Void>`

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java:190]: Another test that would be good to add, would be to have many parallel writers, here you only have one writer.

**Review by ScottDugas — CHANGES_REQUESTED** [review]
Again, this spurred my imagination and I thought of more tests that we could add, and those don't need to be brought into this PR, but would be good to add in the near future.

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java
Comment (jjezra) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java:44]: The test uses `fdb`,  which is defined in `FDBRecordStoreTestBase`.

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java
Comment (jjezra) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java:46]: The ForkJoinPool deadlocks is indeed very concerting, and - FMHO - should be addressed to avoid future issues.
In the context of this test, we have multiple threads that are launched together and do very little in an identical flow, therefore have a **much** higher probability of colliding in that deadlock than the 10 (or more) threads that are triggered by the `indexWriter`. Hence the extra restriction. 
Given enough time, if this deadlock will actually ever happen (let's say - once every 10 years), the merge will be frozen for a while (10 minutes? I'm not sure) and then will get a "no activity" exception (or at least that what've happen in my local environment when it was tested).

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java
Comment (jjezra) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java:91]: ✅

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java
Comment (jjezra) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java:86]: Just for code similarity and testing the nonAgile path. Not critical.

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java
Comment (jjezra) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java:104]: Changed :) 
✅

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java
Comment (jjezra) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java:109]: Calling the commit on the user context forces a flush of agility context. If you wish, I can make it slightly cleaner by calling the commit on the `agilityContext` object. This way the commit will be applied on a "context" what was written.

Comment (jjezra) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java:109]: On a second thought, adding a commit to agilityContext will be just confusing. (Should the nonAgile commit the user context? Should the agile declare itself as closed and forbid future operations? It doesn't seem obvious to the caller). 
Unless you object, I'll leave it this way for now. Or replace `context.commit()` with `agilityContext.flush()`.

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java
Comment (jjezra) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java:136]: ✅

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java
Comment (jjezra) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java:154]: ✅

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java
Comment (jjezra) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java:90]: I've tried to create some arbitrary overlap to ensure no conflicts. But I can change it to your suggestion.

Comment (jjezra) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java:90]: ✅

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java
Comment (jjezra) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java:183]: ✅

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java
Comment (jjezra) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java:182]: ✅

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java
Comment (jjezra) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java:194]: ✅

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java
Comment (jjezra) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java:200]: ✅

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java
Comment (jjezra) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java:204]: 👍 
`accept` and `acceptAsync`?

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java
Comment (jjezra) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java:190]: Good idea! Implemented in both tests (i.e single and multi contexts). 
✅

**Review by ScottDugas — APPROVED** [review]

**Pull request merged**
